`ultralytics 8.3.188` Faster downloads via a single `request`

glenn-jocher opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves download reliability and performance with smarter disk-space checks and modernized type hints; version bump to 8.3.188. 🚀

### 📊 Key Changes
- Bumped version to 8.3.188.
- Modernized type hints across downloads utilities (supports `Path`, unions, tuples; Python 3.11+ style).
- Refactored disk-space check:
  - `check_disk_space` now accepts `file_bytes` instead of a URL, removing a slow HTTP HEAD request.
  - Clearer GB formatting in error messages.
- Smarter download flow in `safe_download`:
  - Disk-space is checked only after getting `Content-Length` from the response (and only for files > 1 MB).
  - `MemoryError` is re-raised immediately (no pointless retries on low disk space).
- Function signature updates:
  - `is_url`, `delete_dsstore`, `zip_directory`, `unzip_file`, `safe_download`, `get_github_assets`, `attempt_download_asset`, and `download` now have stricter, more descriptive types.
  - `safe_download` now returns `Path | str`.

### 🎯 Purpose & Impact
- Faster and more reliable downloads by avoiding unnecessary network calls and checking disk space only when size is known. ⚡
- Better error handling: immediate feedback on insufficient disk space, with clearer messages. 🧯
- Improved developer experience with precise type hints for IDEs and static analysis. 🛠️
- Potential minor breaking change for anyone calling `check_disk_space(url=...)`; update to `check_disk_space(file_bytes=...)`. 📣

glenn-jocher committed:
Remove redundant HEAD request on download checks

UltralyticsAssistant commented:
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist you shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR 🧠
- The `check_disk_space` signature change from `url` to `file_size_bytes` is a breaking internal API change. Please confirm all internal callers are updated accordingly and that no unused imports remain after removing `requests`.
- Since `safe_download` now checks disk space only on the first attempt and only when `Content-Length` is present, please ensure tests cover:
  - `urllib` path with and without `Content-Length`
  - `curl` path (no pre-check) behavior
  - `MemoryError` early exit logic and no-retry pathway
- Consider adding a brief before/after note in the PR description with observed latency improvements or import-time impact (even qualitative) to document the performance gain. ⚡
- If any user-facing behavior or error messaging changed, please update the [documentation](https://docs.ultralytics.com/) accordingly.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Remove redundant HEAD request on download checks

glenn-jocher committed:
Merge remote-tracking branch 'origin/request-once' into request-once

glenn-jocher committed:
Update downloads.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Remove redundant HEAD request on download checks

glenn-jocher committed:
Update downloads.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update __init__.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

Laughing-q commented:
@glenn-jocher what profile tool you are using?

glenn-jocher committed:
Add type hints

glenn-jocher committed:
Merge 597aa2af8a477e555a31f13dee2b4f2780c2ff4f into 52fceb8444c22c480ac3f94d3f4105ca598909ab

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

glenn-jocher commented:
@Laughing-q I just made my own here, but we need something more generic that accepts any python/CLI commands:


```
import cProfile
import pstats
import functools
import inspect
from pathlib import Path
from typing import Dict, List, Callable, Any, Tuple
from ultralytics import YOLO


class YOLOProfiler:
    """Profiles YOLO11 inference and captures function I/O for optimization."""

    def __init__(self):
        self.io_data: Dict[str, List[Dict]] = {}
        self.original_methods = {}

    def capture_io(self, func: Callable) -> Callable:
        """Decorator to capture function inputs and outputs."""

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            # Capture inputs
            input_data = {
                "args_types": [type(arg).__name__ for arg in args],
                "args_shapes": [getattr(arg, "shape", None) for arg in args],
                "kwargs_types": {k: type(v).__name__ for k, v in kwargs.items()},
                "kwargs_shapes": {k: getattr(v, "shape", None) for k, v in kwargs.items()},
            }

            result = func(*args, **kwargs)

            # Capture output
            output_data = {
                "type": type(result).__name__,
                "shape": getattr(result, "shape", None),
                "len": len(result) if hasattr(result, "__len__") else None,
            }

            func_key = f"{func.__module__}.{func.__qualname__}"
            if func_key not in self.io_data:
                self.io_data[func_key] = []

            self.io_data[func_key].append({"input": input_data, "output": output_data})

            return result

        return wrapper

    def patch_methods(self, target_classes: List[str]):
        """Monkey patch methods to capture I/O."""
        import ultralytics

        for class_path in target_classes:
            try:
                parts = class_path.split(".")
                obj = ultralytics
                for part in parts:
                    obj = getattr(obj, part)

                for attr_name in dir(obj):
                    if not attr_name.startswith("_"):
                        attr = getattr(obj, attr_name)
                        if callable(attr) and hasattr(attr, "__self__"):
                            original_key = f"{class_path}.{attr_name}"
                            self.original_methods[original_key] = attr
                            wrapped = self.capture_io(attr)
                            setattr(obj, attr_name, wrapped)
            except AttributeError:
                continue

    def profile_inference(
        self, model_name: str = "yolo11n.pt", source: str = "https://ultralytics.com/images/bus.jpg"
    ) -> Tuple[Dict, Dict]:
        """Profile YOLO11 inference and return timing + I/O data."""
        # Patch key methods before profiling
        target_classes = ["models.yolo.model.YOLO", "nn.modules.head.Detect", "nn.modules.conv.Conv"]
        self.patch_methods(target_classes)

        def run_inference():
            model = YOLO(model_name)
            return model(source)

        # Profile execution
        profiler = cProfile.Profile()
        profiler.enable()
        results = run_inference()
        profiler.disable()

        # Extract ultralytics timing data
        stats = pstats.Stats(profiler)
        timings = {}

        for func_info, (call_count, _, total_time, cumulative_time, _) in stats.stats.items():
            filename, line_num, func_name = func_info

            if "ultralytics" in filename and "profile_tests.py" not in filename:
                # Extract class name from code context if it's a method
                try:
                    import linecache

                    line = linecache.getline(filename, line_num).strip()
                    # Look for class definition by scanning backwards
                    class_name = None
                    for i in range(line_num, max(0, line_num - 100), -1):
                        prev_line = linecache.getline(filename, i).strip()
                        if prev_line.startswith("class ") and ":" in prev_line:
                            class_name = prev_line.split("class ")[1].split("(")[0].split(":")[0].strip()
                            break

                    # Format function name with class if found
                    if class_name and func_name != "<module>" and not func_name.startswith("<"):
                        display_func = f"{class_name}.{func_name}"
                    else:
                        display_func = func_name
                except:
                    display_func = func_name

                key = f"{Path(filename).name}:{display_func}:{line_num}"
                timings[key] = {
                    "cumulative_time": cumulative_time,
                    "total_time": total_time,
                    "call_count": call_count,
                    "avg_time": cumulative_time / call_count if call_count > 0 else 0,
                    "filename": filename,
                }

        return (dict(sorted(timings.items(), key=lambda x: x[1]["cumulative_time"], reverse=True)), self.io_data)

    def analyze_hotspots(self, timings: Dict, io_data: Dict, top_n: int = 10):
        """Analyze and report performance hotspots."""
        print(f"\n=== YOLO11 Inference Hotspots ===")

        # Slowest functions by self time (excluding subfunctions)
        print(f"\n🐌 TOP {top_n} SLOWEST FUNCTIONS (Self Time):")
        print(f"{'Function':<60} {'Self Time':<10} {'Calls':<8} {'Avg Time':<10}")
        print("-" * 90)

        slowest_self = sorted(timings.items(), key=lambda x: x[1]["total_time"], reverse=True)[:top_n]
        for func_key, data in slowest_self:
            avg_self = data["total_time"] / data["call_count"] if data["call_count"] > 0 else 0
            print(f"{func_key:<60} {data['total_time']:<10.4f} {data['call_count']:<8} {avg_self:<10.6f}")

        # Slowest functions by cumulative time (including subfunctions)
        print(f"\n⏱️  TOP {top_n} SLOWEST FUNCTIONS (Cumulative Time):")
        print(f"{'Function':<60} {'Cum Time':<10} {'Self Time':<10} {'Calls':<8} {'Self/Cum':<10}")
        print("-" * 100)

        for i, (func_key, data) in enumerate(timings.items()):
            if i >= top_n:
                break
            ratio = (data["total_time"] / data["cumulative_time"] * 100) if data["cumulative_time"] > 0 else 0
            print(
                f"{func_key:<60} {data['cumulative_time']:<10.4f} {data['total_time']:<10.4f} {data['call_count']:<8} {ratio:<10.1f}%"
            )

        # Most called functions
        most_called = sorted(timings.items(), key=lambda x: x[1]["call_count"], reverse=True)[:top_n]
        print(f"\n🔥 TOP {top_n} MOST CALLED FUNCTIONS:")
        print(f"{'Function':<60} {'Calls':<8} {'Total Time':<10} {'Avg Time':<10}")
        print("-" * 90)

        for func_key, data in most_called:
            print(f"{func_key:<60} {data['call_count']:<8} {data['total_time']:<10.4f} {data['avg_time']:<10.6f}")

        # I/O Analysis
        if io_data:
            print(f"\n📊 I/O SIGNATURES:")
            for func_key, calls in io_data.items():
                if calls:
                    sample_call = calls[0]
                    print(f"\n{func_key} (called {len(calls)} times):")
                    print(f"  Input:  {sample_call['input']['args_types']} {sample_call['input']['args_shapes']}")
                    print(f"  Output: {sample_call['output']['type']} {sample_call['output']['shape']}")

    def get_optimization_targets(self, timings: Dict, min_time: float = 0.01, min_calls: int = 10) -> List[str]:
        """Identify functions worth optimizing."""
        targets = []
        for func_key, data in timings.items():
            if data["cumulative_time"] > min_time or data["call_count"] > min_calls:
                targets.append(func_key)
        return targets


# Convenience function for quick analysis
def profile_yolo_inference(
    model: str = "yolo11n.pt", source: str = "https://ultralytics.com/images/bus.jpg", top_n: int = 10
):
    """Quick YOLO11 inference profiling."""
    profiler = YOLOProfiler()
    timings, io_data = profiler.profile_inference(model, source)
    profiler.analyze_hotspots(timings, io_data, top_n)
    return profiler.get_optimization_targets(timings)


if __name__ == "__main__":
    # Run profiling
    targets = profile_yolo_inference()
    print(f"\n🎯 Optimization targets: {len(targets)} functions identified")
```

glenn-jocher committed:
Merge branch 'main' into request-once

UltralyticsAssistant commented:
Amazing work merged! Huge thanks to @glenn-jocher and @Laughing-q for elevating our downloads flow and developer experience in PR [Improve downloads reliability and types – #21850](https://github.com/ultralytics/ultralytics/pull/21850). 

“Simplicity is the ultimate sophistication.” — Leonardo da Vinci  
Your smarter disk-space checks, modernized type hints, and clear error handling embody that spirit—making downloads faster, more reliable, and easier to reason about. The version bump to 8.3.188 reflects real momentum, and the careful change from `check_disk_space(url=...)` to `check_disk_space(file_bytes=...)` shows thoughtful design that puts users first.

Appreciate the craft and care—this is impact that developers will feel every day. 🚀👏
